File: twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    var hashMap = {};
    for (var i = 0; i < nums.length; i++) {
        /* In JS always save reused variable to improve performance */
        var curr = nums[i];
        var pair = target - curr;
        var match = hashMap[pair];
        if (match === undefined) {
            hashMap[curr] = i + 1;
        } else {
            return match > i + 1 ? [i + 1, match] : [ match, i + 1];
        }

    }
};

File: twoSum.java
public class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i<nums.length; i++) {
            int x = nums[i];
            if (map.containsKey(target-x)) {
                return new int[] {map.get(target-x)+1, i+1};
            } else {
                map.put(x, i);
            }
        }
        throw new IllegalArgumentException("No solution");
    }
}
